diff -uNr ipmitool-1.8.15/lib/ipmi_hpmfwupg.c ipmitool-rflash/lib/ipmi_hpmfwupg.c
--- ipmitool-1.8.15/lib/ipmi_hpmfwupg.c	2014-08-11 13:31:15.000000000 -0400
+++ ipmitool-rflash/lib/ipmi_hpmfwupg.c	2015-11-03 02:39:22.022585519 -0500
@@ -724,18 +724,6 @@
 			lprintf(LOG_NOTICE, "\n    Upgrade undesirable at this moment");
 			return HPMFWUPG_ERROR;
 		}
-		/* Get confimation from the user if he wants to continue when
-		 * service affected during upgrade
-		 */
-		if (!(option & COMPARE_MODE)
-				&& (pFwupgCtx->targetCap.GlobalCapabilities.bitField.servAffectDuringUpg == 1
-					|| pImageHeader->imageCapabilities.bitField.servAffected == 1)) {
-			if (HpmGetUserInput("\nServices may be affected during upgrade. Do you wish to continue? (y/n): ")) {
-				rc = HPMFWUPG_SUCCESS;
-			} else {
-				return HPMFWUPG_ERROR;
-			}
-		}
 	}
 	/* Get the general properties of each component present in image */
 	for (componentId = HPMFWUPG_COMPONENT_ID_0;
