--- ipmitool-1.8.11/lib/ipmi_event.c	2011-04-27 15:25:45.000000000 -0400
+++ ipmitool-1.8.11-eventfix/lib/ipmi_event.c	2011-04-27 15:26:04.000000000 -0400
@@ -97,7 +97,7 @@
 	if (chmed == IPMI_CHANNEL_MEDIUM_SYSTEM) {
 		/* system interface, need extra generator ID */
 		req.msg.data_len = 8;
-		rqdata[0] = 0x20;
+		rqdata[0] = 0x41;
 		memcpy(rqdata+1, emsg, sizeof(struct platform_event_msg));
 	}
 	else {
