diff -ur yaboot-1.3.17-rc1/second/fs_of.c yaboot-1.3.17-rc1-promclaim/second/fs_of.c
--- yaboot-1.3.17-rc1/second/fs_of.c    2011-06-01 22:49:52.036062925 -0400
+++ yaboot-1.3.17-rc1-promclaim/second/fs_of.c  2011-06-22 05:05:03.056063235 -0400
@@ -44,7 +44,7 @@
 #include "errors.h"
 #include "debug.h"

-#define LOAD_BUFFER_SIZE       0x1800000
+#define LOAD_BUFFER_SIZE       0x2000000

 static int of_open(struct boot_file_t* file,
                   struct partition_t* part, struct boot_fspec_t* fspec);
