diff --git a/src/config/branding.h b/src/config/branding.h
index 73f00af..29985fe 100644
--- a/src/config/branding.h
+++ b/src/config/branding.h
@@ -24,8 +24,8 @@ FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
  * "iPXE".
  *
  */
-#define PRODUCT_NAME ""
-#define PRODUCT_SHORT_NAME "iPXE"
+#define PRODUCT_NAME "xCAT Network Boot Agent"
+#define PRODUCT_SHORT_NAME "xNBA"
 #define PRODUCT_URI "http://ipxe.org"
 
 /*
