diff --git a/src/include/ipxe/dhcp.h b/src/include/ipxe/dhcp.h
index b7a5f00..d086105 100644
--- a/src/include/ipxe/dhcp.h
+++ b/src/include/ipxe/dhcp.h
@@ -293,11 +293,11 @@ enum dhcp_client_architecture_values {
 	/** EFI IA32 */
 	DHCP_CLIENT_ARCHITECTURE_IA32 = 0x0006,
 	/** EFI x86-64 */
-	DHCP_CLIENT_ARCHITECTURE_X86_64 = 0x0007,
+	DHCP_CLIENT_ARCHITECTURE_X86_64 = 0x0009, /** swapped with DHCP_CLIENT_ARCHITECTURE_EFI **/
 	/** EFI Xscale */
 	DHCP_CLIENT_ARCHITECTURE_XSCALE = 0x0008,
 	/** EFI BC */
-	DHCP_CLIENT_ARCHITECTURE_EFI = 0x0009,
+	DHCP_CLIENT_ARCHITECTURE_EFI = 0x0007, /** swapped with DHCP_CLIENT_ARCHITECTURE_X86_64 **/
 	/** EFI 32-bit ARM */
 	DHCP_CLIENT_ARCHITECTURE_ARM32 = 0x000a,
 	/** EFI 64-bit ARM */
