diff -urN ipxe/src/config/general.h ipxe-branded/src/config/general.h
--- ipxe/src/config/general.h	2010-11-11 14:09:48.052334980 -0500
+++ ipxe-branded/src/config/general.h	2010-11-12 09:33:37.493584218 -0500
@@ -24,8 +24,8 @@
  * "iPXE".
  *
  */
-#define PRODUCT_NAME ""
-#define PRODUCT_SHORT_NAME "iPXE"
+#define PRODUCT_NAME "xCAT Network Boot Agent"
+#define PRODUCT_SHORT_NAME "xNBA"
 
 /*
  * Timer configuration
