diff -urN gpxe-0.9.9/src/config/general.h gpxe-0.9.9-kvmconfig/src/config/general.h
--- gpxe-0.9.9/src/config/general.h	2009-10-21 15:05:47.000000000 -0400
+++ gpxe-0.9.9-kvmconfig/src/config/general.h	2010-01-26 15:21:58.000000000 -0500
@@ -60,9 +60,9 @@
  *
  */
 
-//#undef	SANBOOT_PROTO_ISCSI	/* iSCSI protocol */
-//#undef	SANBOOT_PROTO_AOE	/* AoE protocol */
-//#undef	SANBOOT_PROTO_IB_SRP	/* Infiniband SCSI RDMA protocol */
+#undef	SANBOOT_PROTO_ISCSI	/* iSCSI protocol */
+#undef	SANBOOT_PROTO_AOE	/* AoE protocol */
+#undef	SANBOOT_PROTO_IB_SRP	/* Infiniband SCSI RDMA protocol */
 
 /*
  * Name resolution modules
@@ -79,34 +79,34 @@
  * you want to use.
  *
  */
-//#define	IMAGE_NBI		/* NBI image support */
-//#define	IMAGE_ELF		/* ELF image support */
-//#define	IMAGE_FREEBSD		/* FreeBSD kernel image support */
-//#define	IMAGE_MULTIBOOT		/* MultiBoot image support */
-//#define	IMAGE_AOUT		/* a.out image support */
-//#define	IMAGE_WINCE		/* WinCE image support */
-//#define	IMAGE_PXE		/* PXE image support */
-//#define	IMAGE_SCRIPT		/* gPXE script image support */
-//#define	IMAGE_BZIMAGE		/* Linux bzImage image support */
-//#define	IMAGE_COMBOOT		/* SYSLINUX COMBOOT image support */
-//#define	IMAGE_EFI		/* EFI image support */
+#undef	IMAGE_NBI		/* NBI image support */
+#undef	IMAGE_ELF		/* ELF image support */
+#undef	IMAGE_FREEBSD		/* FreeBSD kernel image support */
+#undef	IMAGE_MULTIBOOT		/* MultiBoot image support */
+#undef	IMAGE_AOUT		/* a.out image support */
+#undef	IMAGE_WINCE		/* WinCE image support */
+#define	IMAGE_PXE		/* PXE image support */
+#define	IMAGE_SCRIPT		/* gPXE script image support */
+#define	IMAGE_BZIMAGE		/* Linux bzImage image support */
+#undef	IMAGE_COMBOOT		/* SYSLINUX COMBOOT image support */
+#undef	IMAGE_EFI		/* EFI image support */
 
 /*
  * Command-line commands to include
  *
  */
-#define	AUTOBOOT_CMD		/* Automatic booting */
+#undef	AUTOBOOT_CMD		/* Automatic booting */
 #define	NVO_CMD			/* Non-volatile option storage commands */
-#define	CONFIG_CMD		/* Option configuration console */
-#define	IFMGMT_CMD		/* Interface management commands */
-#define	ROUTE_CMD		/* Routing table management commands */
-#define IMAGE_CMD		/* Image management commands */
-#define DHCP_CMD		/* DHCP management commands */
-#define SANBOOT_CMD		/* SAN boot commands */
-#define LOGIN_CMD		/* Login command */
+#undef	CONFIG_CMD		/* Option configuration console */
+#undef	IFMGMT_CMD		/* Interface management commands */
+#undef	ROUTE_CMD		/* Routing table management commands */
+#define IMAGE_CMD		/* Image management commands */
+#undef DHCP_CMD		/* DHCP management commands */
+#undef SANBOOT_CMD		/* SAN boot commands */
+#undef LOGIN_CMD		/* Login command */
 #undef	TIME_CMD		/* Time commands */
 #undef	DIGEST_CMD		/* Image crypto digest commands */
-//#undef	PXE_CMD			/* PXE commands */
+#undef	PXE_CMD			/* PXE commands */
 
 /*
  * Obscure configuration options
