diff -urN ipxe-expandfilename/src/arch/i386/include/bzimage.h ipxe-cmdlinesize/src/arch/i386/include/bzimage.h
--- ipxe-expandfilename/src/arch/i386/include/bzimage.h	2010-11-11 14:09:48.042334268 -0500
+++ ipxe-cmdlinesize/src/arch/i386/include/bzimage.h	2010-11-12 10:47:08.086084309 -0500
@@ -137,6 +137,6 @@
 #define BZI_STACK_SIZE 0x1000
 
 /** Maximum size of command line */
-#define BZI_CMDLINE_SIZE 0x100
+#define BZI_CMDLINE_SIZE 0x7FF
 
 #endif /* _BZIMAGE_H */
