--- elilo/Make.defaults.orig	2012-05-14 22:29:47.868436029 +0100
+++ elilo/Make.defaults	2012-05-14 22:30:05.968199364 +0100
@@ -55,9 +55,9 @@
 # They are installed as part of the GNU-EFI package installation
 #
 EFIINC	   = /usr/include/efi
-GNUEFILIB  = /usr/lib64
-EFILIB	   = /usr/lib64/gnuefi
-EFICRT0	   = /usr/lib64/gnuefi
+GNUEFILIB  = /usr/lib
+EFILIB	   = /usr/lib
+EFICRT0	   = /usr/lib
 
 CDIR    := $(shell if [ "$$PWD" != "" ]; then echo $$PWD; else pwd; fi)
 TOPDIR  =
