*** net-snmp-5.7.2/configure.orig	Wed Apr 10 09:43:23 2013
--- net-snmp-5.7.2/configure	Wed Apr 10 09:51:13 2013
***************
*** 8845,8853 ****
  # whether `pass_all' will *always* work, you probably want this one.
  
  case $host_os in
! aix[4-9]*)
!   lt_cv_deplibs_check_method=pass_all
!   ;;
  
  beos*)
    lt_cv_deplibs_check_method=pass_all
--- 8845,8853 ----
  # whether `pass_all' will *always* work, you probably want this one.
  
  case $host_os in
! aix*)
!    lt_cv_deplibs_check_method=pass_all
!    ;;
  
  beos*)
    lt_cv_deplibs_check_method=pass_all
***************
*** 12420,12426 ****
  
      # See if GNU ld supports shared libraries.
      case $host_os in
!     aix[3-9]*)
        # On AIX/PPC, the GNU linker is very broken
        if test "$host_cpu" != ia64; then
  	ld_shlibs=no
--- 12420,12426 ----
  
      # See if GNU ld supports shared libraries.
      case $host_os in
!     aix*)
        # On AIX/PPC, the GNU linker is very broken
        if test "$host_cpu" != ia64; then
  	ld_shlibs=no
***************
*** 12687,12693 ****
        fi
        ;;
  
!     aix[4-9]*)
        if test "$host_cpu" = ia64; then
  	# On IA64, the linker does run time linking by default, so we don't
  	# have to do anything special.
--- 12687,12693 ----
        fi
        ;;
  
!     aix*)
        if test "$host_cpu" = ia64; then
  	# On IA64, the linker does run time linking by default, so we don't
  	# have to do anything special.
***************
*** 12709,12715 ****
  	# Test if we are trying to use run time linking or normal
  	# AIX style linking. If -brtl is somewhere in LDFLAGS, we
  	# need to do runtime linking.
! 	case $host_os in aix4.[23]|aix4.[23].*|aix[5-9]*)
  	  for ld_flag in $LDFLAGS; do
  	  if (test $ld_flag = "-brtl" || test $ld_flag = "-Wl,-brtl"); then
  	    aix_use_runtimelinking=yes
--- 12709,12715 ----
  	# Test if we are trying to use run time linking or normal
  	# AIX style linking. If -brtl is somewhere in LDFLAGS, we
  	# need to do runtime linking.
! 	case $host_os in aix4.[23]|aix4.[23].*|aix[5-9]* | aix*)
  	  for ld_flag in $LDFLAGS; do
  	  if (test $ld_flag = "-brtl" || test $ld_flag = "-Wl,-brtl"); then
  	    aix_use_runtimelinking=yes
***************
*** 13803,13809 ****
    soname_spec='${libname}${release}${shared_ext}$major'
    ;;
  
! aix[4-9]*)
    version_type=linux
    need_lib_prefix=no
    need_version=no
--- 13803,13809 ----
    soname_spec='${libname}${release}${shared_ext}$major'
    ;;
  
! aix*)
    version_type=linux
    need_lib_prefix=no
    need_version=no
***************
*** 15174,15180 ****
      fi
      ;;
  
!   aix[4-9]*)
      if test "$host_cpu" != ia64 && test "$aix_use_runtimelinking" = no ; then
        test "$enable_shared" = yes && enable_static=no
      fi
--- 15174,15180 ----
      fi
      ;;
  
!   aix*)
      if test "$host_cpu" != ia64 && test "$aix_use_runtimelinking" = no ; then
        test "$enable_shared" = yes && enable_static=no
      fi
***************
*** 15630,15636 ****
  #
  
  case $target_os in
!     aix4.2*|aix4.3*|aix5*|aix6*|aix7*)
          #   Runtime linking flags                       (AIX)
          #
          { $as_echo "$as_me:${as_lineno-$LINENO}: checking what linker flags to add to enable runtime linking on AIX" >&5
--- 15630,15636 ----
  #
  
  case $target_os in
!     aix4.2*|aix4.3*|aix5*|aix6*|aix7*|aix*)
          #   Runtime linking flags                       (AIX)
          #
          { $as_echo "$as_me:${as_lineno-$LINENO}: checking what linker flags to add to enable runtime linking on AIX" >&5
